### PR TITLE
Email notification on project registration/cancellation

### PR DESCRIPTION
**Pull request by qurbat**
This pull request contains code for sending email confirmations once a user has registered or de-registered for a project. Upon registering, the dispatched email includes an `.ics` calendar file and details about the project, including project title, schedule start date, and project URL. Upon de-registering for a project, an email is dispatched to the user confirming the same.

### CODE REVIEW

File: funnel/signals.py [hunk 1]
```diff
@@ -45,6 +45,7 @@
 org_data_changed = app_signals.signal('org-data-changed')
 team_data_changed = app_signals.signal('team-data-changed')
 session_revoked = app_signals.signal('session-revoked')
+project_email_dispatched = app_signals.signal('project_email_dispatched')
```
Comment (jace) [inline funnel/signals.py:48]: Incorrect name. No email has been dispatched at the time this signal is raised. This signal indicates that a user has registered for a project, so name it appropriately.

File: funnel/views/project.py [hunk 2]
```diff
@@ -572,6 +573,9 @@ def register(self):
                 rsvp.rsvp_yes()
                 db.session.commit()
                 flash(_("You have successfully registered"), 'success')
+                send_email_for_project_registration(
```
Comment (jace) [inline funnel/views/project.py:576]: Send the signal here. Don't call the email function directly. The signal's sender is the `rsvp` object that was just created.

Comment (qurbat) [reply funnel/views/project.py]: I made changes. Does [this](https://github.com/hasgeek/funnel/blob/5fdc63d90cf3a9cf5d28664ca14058a483368370/funnel/views/project.py) look OK?

File: funnel/templates/email_project_registration.html.jinja2 [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{% extends "email_layout.html.jinja2" %}
+{% block content -%}
+
+<p>Hi {{ current_auth.user.fullname }},</p>
+
+<p>Your registration for <b>{{ project.title }}</b> was successful.</b></p>
```
Comment (jace) [inline funnel/templates/email_project_registration.html.jinja2:6]: Second `</b>` is extra

File: funnel/templates/email_project_registration.html.jinja2 [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+<p>The project is scheduled to begin at {{ project.schedule_start_at_localized|datetime }}</p>
```
Comment (jace) [inline funnel/templates/email_project_registration.html.jinja2:8]: For a multi-day project, this could be in the past. Use the next upcoming session time instead. Also, sessions have a start time, projects don't. Your text should be "The next session is at..."

File: funnel/templates/email_project_registration.html.jinja2 [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+<p>More details about the project are available here: {{ project.url_for(_external=True) }}</p>
```
Comment (jace) [inline funnel/templates/email_project_registration.html.jinja2:10]: Every sentence in this email needs a `{% trans %}` tag. Further, since this is being sent in a background job, you don't have the user's locale from the current request. The locale must be captured during the request (use `get_locale()` from Baseframe) and passed as a parameter to the background job.

File: funnel/views/email.py [hunk 6]
```diff
@@ -39,7 +40,7 @@ def jsonld_confirm_action(description, url, title):
     }
 
 
-def send_email(subject, to, content):
+def send_email(subject, to, content, attachment=[]):
```
Comment (jace) [inline funnel/views/email.py:43]: `attachments`. Plural. *Never* use `[]` or any mutable structure as a default parameter or memory leaks will happen. Use `()` or `None`.

File: funnel/views/email.py [hunk 7]
```diff
@@ -57,7 +58,8 @@ def send_email(subject, to, content):
     msg = EmailMultiAlternatives(
         subject=subject, to=to, body=body, alternatives=[(html, 'text/html')]
     )
-
+    if attachment != []:
```
Comment (jace) [inline funnel/views/email.py:61]: WTF is this? The structure you need:

```python
if attachments:
    for attachment in attachments:
        # Do whatever
```

File: funnel/views/email.py [hunk 8]
```diff
@@ -57,7 +58,8 @@ def send_email(subject, to, content):
...
+        msg.attach(content=attachment, filename='invite.ics', mimetype='text/calendar')
```
Comment (jace) [inline funnel/views/email.py:62]: Why is the filename and mimetype hardcoded? These should be parameters.

File: funnel/views/email.py [hunk 9]
```diff
@@ -139,6 +141,20 @@ def send_email_for_organization_admin_membership_revoked(
...
 
 
+@signals.user_registered_for_project.connect
+def send_email_for_project_registration(sender, project, user):
+    send_email(
+        subject=_("Registration confirmation for {project}").format(
+            project=project.title
+        ),
+        to=[user],
+        attachment=schedule_ical(project),
```
Comment (jace) [inline funnel/views/email.py:151]: This should be a background job. `schedule_ical` is not a cheap call.

File: funnel/views/project.py [hunk 10]
```diff
@@ -572,6 +573,9 @@ def register(self):
                 rsvp.rsvp_yes()
                 db.session.commit()
                 flash(_("You have successfully registered"), 'success')
+                user_registered_for_project.send(
+                    project=self.obj, user=current_auth.user
+                )
```
Comment (jace) [inline funnel/views/project.py:578]: Signals require a positional parameter for the sender. This call should be `signal_name.send(self.obj, user=current_auth.user)`

File: funnel/views/schedule.py [hunk 11]
```diff
@@ -226,7 +244,7 @@ def schedule_json(self):
 
     @route('ical')
     @requires_roles({'reader'})
-    def schedule_ical(self):
+    def schedule_ical_download(self):
```
Comment (jace) [inline funnel/views/schedule.py:247]: Why is this a dupe of the function above? One function can call the other function.

File: funnel/views/email.py (hunk 6 above)
Comment (jace) [reply funnel/views/email.py:43]: https://deepsource.io/gh/hasgeek/funnel/run/29eea7df-50ae-461b-9d96-ba904dae7413/python/PYL-W0102

File: funnel/templates/email_project_registration.html.jinja2 [hunk 12]
```diff
@@ -0,0 +1,12 @@
+{% extends "email_layout.html.jinja2" %}
+{% block content -%}
+
+<p>{% trans %}Hi{% endtrans %} {{ user.fullname }},</p>
```
Comment (jace) [inline funnel/templates/email_project_registration.html.jinja2:4]: `{% trans user=user.fullname %}Hi {{user}},{% endtrans %}`

This will allow translations where the "Hi" can appear after the name.

File: funnel/templates/email_project_registration.html.jinja2 [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+
+<p>{% trans %}Your registration has been recorded successfully for{% endtrans %} <b>{{ project.title }}</b>.</p>
+
+<p>{% trans %}The next session in the schedule will begin at{% endtrans %} {{ project.next_session_at|datetime }}</p>
+
+<p>{% trans %}More details about the project are available at this link{% endtrans %}: {{ project.url_for(_external=True) }}</p>
```
Comment (jace) [inline funnel/templates/email_project_registration.html.jinja2:10]: Same for each of these. Don't cut sentences.

File: funnel/views/email.py [hunk 14]
```diff
@@ -39,7 +41,7 @@ def jsonld_confirm_action(description, url, title):
     }
 
 
-def send_email(subject, to, content):
+def send_email(subject, to, content, attachments=None, filename=None, mimetype=None):
```
Comment (jace) [inline funnel/views/email.py:44]: `filename` and `mimetype` are unused parameters.

File: funnel/views/email.py [hunk 15]
```diff
@@ -53,11 +55,17 @@ def send_email(subject, to, content):
         for recipient in to
     ]
     body = html2text(content)
-    html = transform(content, base_url=request.url_root)
+    html = transform(content, base_url=app.config['DEFAULT_DOMAIN'])
```
Comment (jace) [inline funnel/views/email.py:58]: `base_url` must be a URL, not a domain. You can use `f'https://{app.config['DEFAULT_DOMAIN']}/'` but this will probably need something more robust down the road.

File: funnel/views/email.py [hunk 16]
```diff
@@ -139,6 +147,37 @@ def send_email_for_organization_admin_membership_revoked(
...
+@signals.user_registered_for_project.connect
+def send_email_for_project_registration(project, user, locale=None):
+    background_project_email.queue(
+        project_id=project.id, user_id=user.id, locale=get_locale()
+    )
+
+
+@rq.job('funnel')
+def background_project_email(project_id, user_id, locale=None):
+    EmailAttachment = namedtuple('EmailAttachment', ['file', 'filename', 'mimetype'])
```
Comment (jace) [inline funnel/views/email.py:159]: `namedtuple` creates classes. The camel casing of `EmailAttachment` is the hint. Create this class at the top of this file, right after imports.

File: funnel/views/email.py [hunk 17]
```diff
@@ -139,6 +147,37 @@ def send_email_for_organization_admin_membership_revoked(
...
 
+@signals.user_registered_for_project.connect
+def send_email_for_project_registration(project, user, locale=None):
+    background_project_email.queue(
+        project_id=project.id, user_id=user.id, locale=get_locale()
+    )
+
+
+@rq.job('funnel')
+def background_project_email(project_id, user_id, locale=None):
```
Comment (jace) [inline funnel/views/email.py:158]: Locale must be supplied. Don't take `=None`.

File: funnel/views/email.py [hunk 18]
```diff
@@ -139,6 +147,37 @@ def send_email_for_organization_admin_membership_revoked(
...
+    EmailAttachment = namedtuple('EmailAttachment', ['file', 'filename', 'mimetype'])
+    project = Project.query.get(project_id)
+    user = User.query.get(user_id)
+    with app.app_context():
+        send_email(
```
Comment (jace) [inline funnel/views/email.py:163]: You have to set the locale first.

```python
from flask_babelhg import force_locale
…
with force_locale(locale):
    send_email(…)
```

File: funnel/views/project.py [hunk 19]
```diff
@@ -577,6 +578,9 @@ def register(self):
                 rsvp.rsvp_yes()
                 db.session.commit()
                 flash(_("You have successfully registered"), 'success')
+                user_registered_for_project.send(
+                    self.obj, user=current_auth.user, locale=get_locale()
```
Comment (jace) [inline funnel/views/project.py:582]: Move this `get_locale` call to the signal receiver. It's not the view's problem. The signal receiver is the one triggering a background job that needs the locale.

File: funnel/views/email.py [hunk 20]
```diff
@@ -139,6 +147,37 @@ def send_email_for_organization_admin_membership_revoked(
...
+            subject=_("Registration confirmation for {project}").format(
+                project=project.title
+            ),
+            to=[user],
+            attachments=[
+                EmailAttachment(
+                    file=schedule_ical(project),
```
Comment (jace) [inline funnel/views/email.py:170]: Rename this field to `content`. Use `file` only if it's a file-like object. Over here you have a string.

File: funnel/views/email.py [hunk 21]
```diff
@@ -1,15 +1,20 @@
...
 
 from flask import current_app, render_template, request, url_for
 from flask_mailman import EmailMultiAlternatives
 
+from flask_babelhg import force_locale
 from html2text import html2text
 from premailer import transform
 
-from baseframe import _
+from baseframe import _, get_locale
```
Comment (jace) [inline funnel/views/email.py:11]: I was looking through Flask-Babelhg's code. Try importing this from `flask_babelhg`. It has a bit more of a safety harness.

File: funnel/views/email.py [hunk 22]
```diff
@@ -53,11 +58,18 @@ def send_email(subject, to, content):
         for recipient in to
     ]
     body = html2text(content)
-    html = transform(content, base_url=request.url_root)
+    html = transform(
+        content, base_url=f'https://{app.config["DEFAULT_DOMAIN"]}/')  # NOQA
```
Comment (jace) [inline funnel/views/email.py:62]: Never use naked `# NOQA`. Specify the code you're ignoring. Also, why does this line have a QA problem?

File: funnel/views/email.py [hunk 23]
```diff
@@ -139,6 +151,38 @@ def send_email_for_organization_admin_membership_revoked(
...
+def send_email_for_project_registration(project, user):
+    background_project_email.queue(
+        project_id=project.id, user_id=user.id, locale=get_locale()
+    )
+
+
+@rq.job('funnel')
+def background_project_email(project_id, user_id, locale):
+    project = Project.query.get(project_id)
+    user = User.query.get(user_id)
```
Comment (jace) [inline funnel/views/email.py:164]: Move these two inside the `with`. It doesn't matter here, but only because `__init__.py` has this line: `db.app = app`, telling the database engine to look for config in `app`. If we remove that line for any reason, these lines will fail. Moving inside `app_context` will resolve that.

File: funnel/views/project.py [hunk 24]
```diff
@@ -196,12 +197,7 @@ def feature_project_has_no_sessions(obj):
 
 @Project.features('comment_new')
 def feature_project_comment_new(obj):
-    return obj.current_roles.participant
-
-
-@Project.features('post_update')
-def feature_project_post_update(obj):
-    return obj.current_roles.editor
+    return obj.current_roles.participant is True
```
Comment (jace) [inline funnel/views/project.py:200]: Why is this change here? It's not related.

File: funnel/views/project.py [hunk 25]
```diff
@@ -271,7 +267,7 @@ def view(self):
         rsvp_form = RsvpTransitionForm()
         current_rsvp = self.obj.rsvp_for(current_auth.user)
         return {
-            'project': self.obj.current_access(),
+            'project': self.obj,
```
Comment (jace) [inline funnel/views/project.py:270]: This too.

File: funnel/views/project.py [hunk 26]
```diff
@@ -775,6 +772,8 @@ def toggle_featured(self):
         if featured_form.validate_on_submit():
             self.obj.featured = not self.obj.featured
             db.session.commit()
+            if self.obj.featured:
+                flash(_("Your project is now a spotlight on homepage"), 'info')
```
Comment (jace) [inline funnel/views/project.py:776]: ?

File: funnel/views/project.py [hunk 27]
```diff
@@ -455,7 +451,7 @@ def edit_banner(self):
                 form.populate_obj(self.obj)
                 db.session.commit()
                 flash(_("Your changes have been saved"), 'info')
-                return redirect(self.obj.url_for(), code=303)
+                return render_redirect(self.obj.url_for(), code=303)
```
Comment (jace) [inline funnel/views/project.py:454]: Unrelated change

File: funnel/templates/email_project_registration.html.jinja2 [hunk 28]
```diff
@@ -0,0 +1,12 @@
+{% extends "email_layout.html.jinja2" %}
+{% block content -%}
+
+<p>{% trans user=user.fullname %}Hi {{user}},{% endtrans %}</p>
+
+<p>{% trans project=project.title%}Your registration for the project <b>'{{ project }}'</b> has been recorded successfully.{% endtrans %}</p>
```
Comment (jace) [inline funnel/templates/email_project_registration.html.jinja2:6]: This line needs better language.

File: funnel/views/email.py [hunk 29]
```diff
@@ -139,6 +150,51 @@ def send_email_for_organization_admin_membership_revoked(
...
+def send_email_for_project_registration(project, user):
+    background_project_registration_email.queue(
+        project_id=project.id, user_id=user.id, locale=get_locale()
+    )
+
+
+@signals.user_cancelled_project_registration.connect
+def send_email_for_project_deregistration(project, user):
+    locale = get_locale()
+    with force_locale(locale):
```
Comment (jace) [inline funnel/views/email.py:163]: This is a noop. You're already in this locale. You can remove both lines.

File: funnel/views/email.py [hunk 30]
```diff
@@ -139,6 +150,51 @@ def send_email_for_organization_admin_membership_revoked(
...
 
+@signals.user_registered_for_project.connect
+def send_email_for_project_registration(project, user):
+    background_project_registration_email.queue(
+        project_id=project.id, user_id=user.id, locale=get_locale()
+    )
+
+
+@signals.user_cancelled_project_registration.connect
+def send_email_for_project_deregistration(project, user):
```
Comment (jace) [inline funnel/views/email.py:161]: The first parameter to both signals should be the `rsvp` object, not the project. Send project as a separate parameter.

File: funnel/views/project.py [hunk 31]
```diff
@@ -591,6 +593,9 @@ def deregister(self):
                 rsvp.rsvp_no()
                 db.session.commit()
                 flash(_("Your registration has been cancelled"), 'info')
+                user_cancelled_project_registration.send(
+                    self.obj, user=current_auth.user
+                )
```
Comment (jace) [inline funnel/views/project.py:598]: Same.

File: funnel/templates/email_project_registration.html.jinja2 [hunk 32]
```diff
@@ -0,0 +1,12 @@
+{% extends "email_layout.html.jinja2" %}
+{% block content -%}
+
+<p>{% trans user=user.fullname %}Hi {{ user }},{% endtrans %}</p>
+
+<p>{% trans project=project.title %}You have successfully registered for <b>{{ project }}</b>.{% endtrans %}</p>
+
+<p>{% trans next_session=project.next_session_at|datetime %}The next session in the schedule will begin at {{ next_session }}{% endtrans %}</p>
+
+<p>{% trans project_url=project.url_for(_external=True) %}More details about the project are available at this link: {{ project_url }}{% endtrans %}</p>
```
Comment (jace) [inline funnel/templates/email_project_registration.html.jinja2:10]: `<p><a class="btn" href="{{ project.url_for(_external=true) }}">{% trans %}Project page{% endtrans %}</a></p>`

(Jinja2 uses lowercase `true`, `false` and `none`.)

File: funnel/templates/email_project_deregister.html.jinja2 [hunk 33]
```diff
@@ -0,0 +1,9 @@
+{% extends "email_layout.html.jinja2" %}
+{% block content -%}
+
+<p>{% trans user=user.fullname %}Hi {{ user }},{% endtrans %}</p>
+
+<p>{% trans project=project.title %}You have deregistered for <b>{{ project }}</b>.{% endtrans %}</p>
+
+<p>{% trans project_url=project.url_for(_external=true) %}In case this was an accident, or if you wish to register again, you can do so through this link: {{ project_url }}{% endtrans %}</p>
```
Comment (jace) [inline funnel/templates/email_project_deregister.html.jinja2:8]: Make this a two paragraphs, with a button on the second para.

File: funnel/templates/email_project_registration.html.jinja2 [hunk 34]
```diff
@@ -0,0 +1,12 @@
+{% extends "email_layout.html.jinja2" %}
+{% block content -%}
+
+<p>{% trans user=user.fullname %}Hi {{ user }},{% endtrans %}</p>
+
+<p>{% trans project=project.title %}You have successfully registered for <b>{{ project }}</b>.{% endtrans %}</p>
+
+<p>{% trans next_session=project.next_session_at|datetime %}The next session in the schedule will begin at {{ next_session }}{% endtrans %}</p>
```
Comment (jace) [inline funnel/templates/email_project_registration.html.jinja2:8]: Full stop missing.
